### PR TITLE
HPCC-27612 update placements process order

### PR DESCRIPTION
**Pull request by xwang2713**
No matter what a user input the placements will be process as
"all", "type", "target" and "pod" or "job".

This will make it easier for users since they don't need worry about the order of placements.

 Signed-off-by: Xiaoming Wang <xiaoming.wang@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:

Tested with helm template and verify the content

### CODE REVIEW

**Comment by xwang2713** [comment]
@g-pan just review the hpcc/docs/placements.md.

File: helm/hpcc/docs/placements.md [hunk 1]
```diff
@@ -20,6 +20,8 @@ The list item in "pods" can be one of the following:
 4) Job name regular expression:  For example "compile-" or "compile-.*" or exact match "^compile-.*$"
 5) set pod: ["all"] for all HPCC Systems components
 
+No mattern what order a user inputs the placements will be processed as "all", "type", "target" and "pod" or "job".
```
Comment (jakesmith) [inline helm/hpcc/docs/placements.md:23]: typo: mattern->matter

Comment (jakesmith) [reply helm/hpcc/docs/placements.md:23]: I'd reword as:

Regardless of the order of placements in the configuration, they will be processed in the following order:  "all", "type", "target" and "pod" or "job".

File: helm/hpcc/templates/_helpers.tpl [hunk 2]
```diff
@@ -1334,18 +1334,23 @@ Pass in dict with root, job, target and type
 {{- $job := .job -}}
 {{- $target := (printf "target:%s" .target | default "") -}}
 {{- $type := printf "type:%s" .type -}}
+{{- $categories := list "all" $type $target -}}
 {{- $placementsDict := dict -}}
-{{- range $placement := .root.Values.placements -}}
-{{- if or (has $target $placement.pods) (has $type $placement.pods) (has "all" $placement.pods) -}}
+{{- $placements := .root.Values.placements -}}
+{{- range $category := $categories -}}
+{{- range $placement := $placements -}}
```
Comment (jakesmith) [inline helm/hpcc/templates/_helpers.tpl:1341]: not entirely new, but it is clearer/easy to read if there is block indentation in these template definitions.

File: helm/hpcc/templates/_helpers.tpl [hunk 3]
```diff
@@ -1359,12 +1364,16 @@ Pass in dict with root, pod, target and type
 {{- $pod := .pod -}}
 {{- $target := (printf "target:%s" .target | default "") -}}
 {{- $type := printf "type:%s" .type -}}
+{{- $categories := list "all" $type $target $pod -}}
 {{- $placementsDict := dict  -}}
-{{- range $placement := .root.Values.placements -}}
-{{- if or (has $pod $placement.pods) (has $target $placement.pods) (has $type $placement.pods) (has "all"  $placement.pods) -}}
+{{- $placements := .root.Values.placements -}}
+{{- range $category := $categories -}}
```
Comment (jakesmith) [inline helm/hpcc/templates/_helpers.tpl:1370]: as per last comment - re. block indendation.

**Review by jakesmith — COMMENTED** [review]
@xwang2713 - looks ok, please see comments

File: helm/hpcc/docs/placements.md (hunk 1 above)
Comment (xwang2713) [reply helm/hpcc/docs/placements.md:23]: Fixed

File: helm/hpcc/templates/_helpers.tpl (hunk 2 above)
Comment (xwang2713) [reply helm/hpcc/templates/_helpers.tpl:1341]: Done

File: helm/hpcc/templates/_helpers.tpl (hunk 3 above)
Comment (xwang2713) [reply helm/hpcc/templates/_helpers.tpl:1370]: done

File: helm/hpcc/docs/placements.md [hunk 4]
```diff
@@ -130,7 +132,7 @@ placements:
       effect: "NoSchedule"
 
 ```
-"topologySpreadConstraints" example, there are two node pools which have "hpcc=spot1" and "hpcc=spot2" respectively. The roxie pods will be evenly scheduled on the two node pools. After deployment verify it with
+"topologySpreadConstraints" example, there are two node pools which have "hpcc=nodepool1" and "hpcc=nodepool2" respectively. The roxie pods will be evenly scheduled on the two node pools. After deployment verify it with
```
Comment (g-pan) [inline helm/hpcc/docs/placements.md:135]: Roxie (a proper noun needs initial cap)

**Review by g-pan — COMMENTED** [review]
one minor comment - capitalize Roxie

**Review by jakesmith — APPROVED** [review]
@xwang2713 - looks fine, please merge.

File: helm/hpcc/docs/placements.md (hunk 4 above)
Comment (xwang2713) [reply helm/hpcc/docs/placements.md:135]: Fixed
